### PR TITLE
Add max_output_tokens to OpenAI models and integrate into requests

### PR DESCRIPTION
**Pull request by Cupnfish**
### Pull Request Title
Introduce `max_output_tokens` Field for OpenAI Models

https://platform.deepseek.com/api-docs/news/news0725/#4-8k-max_tokens-betarelease-longer-possibilities

### Description
This commit introduces a new field `max_output_tokens` to the OpenAI models, which allows specifying the maximum number of tokens that can be generated in the output. This field is now integrated into the request handling across multiple crates, ensuring that the output token limit is respected during language model completions.

Changes include:
- Adding `max_output_tokens` to the `Custom` variant of the `open_ai::Model` enum.
- Updating the `into_open_ai` method in `LanguageModelRequest` to accept and use `max_output_tokens`.
- Modifying the `OpenAiLanguageModel` and `CloudLanguageModel` implementations to pass `max_output_tokens` when converting requests.
- Ensuring that the `max_output_tokens` field is correctly serialized and deserialized in relevant structures.

This enhancement provides more control over the output length of OpenAI model responses, improving the flexibility and accuracy of language model interactions.

### Changes
- Added `max_output_tokens` to the `Custom` variant of the `open_ai::Model` enum.
- Updated the `into_open_ai` method in `LanguageModelRequest` to accept and use `max_output_tokens`.
- Modified the `OpenAiLanguageModel` and `CloudLanguageModel` implementations to pass `max_output_tokens` when converting requests.
- Ensured that the `max_output_tokens` field is correctly serialized and deserialized in relevant structures.

### Related Issue
https://github.com/zed-industries/zed/pull/16358

### Screenshots / Media
N/A

### Checklist
- [x] Code compiles correctly.
- [x] All tests pass.
- [ ] Documentation has been updated accordingly.
- [ ] Additional tests have been added to cover new functionality.
- [ ] Relevant documentation has been updated or added.

### Release Notes

- Added `max_output_tokens` field to OpenAI models for controlling output token length.

### CODE REVIEW

**Comment by Cupnfish** [comment]
@cla-bot check

File: crates/open_ai/src/open_ai.rs [hunk 1]
```diff
@@ -113,6 +117,15 @@ impl Model {
             Self::Custom { max_tokens, .. } => *max_tokens,
         }
     }
+
+    pub fn max_output_tokens(&self) -> Option<u32> {
+        match self {
+            Self::Custom {
+                max_output_tokens, ..
+            } => *max_output_tokens,
+            _ => None,
```
Comment (jvmncs) [inline crates/open_ai/src/open_ai.rs:126]: any chance you can include the max output tokens for the other models here? they can be found in the lists on this page: https://platform.openai.com/docs/models

Comment (Cupnfish) [reply crates/open_ai/src/open_ai.rs]: Thank you for the suggestion! I have added support for the max_output_tokens for the relevant models as per the documentation. However, I don’t have the OpenAI tokens to conduct testing myself. If someone could help with the testing, I would greatly appreciate it!

**Review by jvmncs — APPROVED** [review]
lgtm, thanks! looks like this PR conflicts with main now, will merge when that's resolved

**Comment by notpeter** [comment]
FYI. This broke GPT-4 (and only GPT-4). 
image
Fix in:
- https://github.com/zed-industries/zed/pull/17168

**Pull request merged**